src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -94,6 +102,90 @@ pub struct Constraint {
     span: Span,
 }
 
+/// A "type test" corresponds to an outlives constraint between a type
+/// and a lifetime, like `T: 'x` or `<T as Foo>::Bar: 'x`.  They are
+/// translated from the `Verify` region constraints in the ordinary
+/// inference context.
+///
+/// These sorts of constraints are handled differently than ordinary
+/// constraints, at least at present. During type checking, the
+/// `InferCtxt::process_registered_region_obligations` method will
+/// attempt to convert a type test like `T: 'x` into an ordinary
+/// outlives constraint when possible (for example, `&'a T: 'b` will
+/// be converted into `'a: 'b` and registered as a `Constraint`).
+///
+/// In some cases, however, there are outlives relationships that are
+/// not converted into a region constraint, but rather into one of
+/// these "type tests".  The distinction is that a type test does not
+/// influence the inference result, but instead just examines the
+/// values that we ultimately inferred for each region variable and
+/// checks that they meet certain extra criteria.  If not, an error
+/// can be issued.
+///
+/// One reason for this is that these type tests always boil down to a
```

Comment:
This is not *quite* correct - consider `<&'α T as Foo>::Bar: 'x` where `'α` is an inference variable. IIRC at least AST borrowck will convert this to `'α: 'x`

Reply:
Yes. When I was writing this, I thought I remembered that we always issued hard region constraints in that case (over-constrained). But I remembered later that we do not. Still not crazy about this behavior, but the right fix is a pain of course. =) (For later...)